File: problem041.cpp
//Kevin Yeap
/*========== Project Euler: Problem  ==========
We shall say that an n-digit number is pandigital if it makes use of all the 
digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is also prime.

What is the largest n-digit pandigital prime that exists?
=========================*/

/*========== program design notes ==========
ways to approach this problem.

1.
find all permutations of digits 1 - n, 9 being 9,8,7, etc... until the largest pantigital prime is found.   O(n*n!)
brute force prime checking complexity   O(n(sqrt(n)));

2.
find all pandigital permutations    O(n*n!)
find all primes from 1 - 98765321   O(n(sqrt(n)))
sort both lists and find biggest common.

3.
find all primes from 1 - 987654321.  O(n(sqrt(n)))
check pandigital starting from largest prime until found. O(n)


Solution implementation.
Sieve of Eratosthenes to find primes, then check if they are panditigal.
run time complexity
O(Nlog log N) +  O(n)

space complexity O(n)  Acceptable since the number of primes are small.
=========================*/

/*========== program output ==========
        == Project Euler: Problem 41 ==

answer: 
=========================*/

#include <iostream>
#include <vector>
using namespace std;

vector<int> primes(int);

#define MAX 100//987654321

int main() {

  vector<int> primez = primes(MAX);
  
  int result = 0;
  
  cout << "\t== Project Euler: Problem 41 ==" << endl << endl; 
  cout << "answer: " << result << endl << endl; 
  
  return 0;
}


//returns all primes from 1 - n ordered from biggest to smallest.
vector<int> primes(max_prime) {

  //vector of bools is said to use 1 bit per bool. Could also use bitset.
  vector<bool> b_prime(max_prime+1, 0);
  
  //sieve is modified to go through only odds.
  for (int i = 3; i <= max_prime; i+=2) {
    b_prime[i] = true;
  }
  
  for (int i = 3; i < sqrt(max_prime)+1; i+=2) {
    if (b_prime[i]) {
      for (int j = 3*i; j < max_prime+1; j+=2*i) {
        b_prime[j] = false; 
      }   
    }
  }
  
  vector<int> primes;
  
  int index = b_prime.length();
  while (index >= 0) {
    if (b_prime[index]) {
      primes.push_back(2*index+3);
      cout << 2*index+3 << endl; 
    }
    index--;  
  }
  
  primes.push_back(2);

  return primes;
}

File: problem041.go
//Kevin Yeap
/*========== Project Euler: Problem 41 ==========
We shall say that an n-digit number is pandigital if it makes use of all the
digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is
also prime.

What is the largest n-digit pandigital prime that exists?
=========================*/

/*========== program design notes ==========
answer is a prime. means that the number is odd.

ways to approach this.
loop through all pandigitals, and filter out primes.
loop through all primes and filter out pandigitals.

it would probably be a better idea to find the primes first since finding
those will also filter out even pandigitals.

sieve of eratosthenes to find primes.
check pandigital starting from the largest prime.

WARNING: since. Go represents a bool in a byte (not a bit).
so with an array of 7654321 bools,this means that the amount of free memory
required to run this program without a runtime error is about 1 GB.

using divisible by 3 rule we can rule out all 8 and 9 digit numbers
=========================*/

/*========== program output ==========
        == Project Euler: Problem 41 ==

answer:7652413
=========================*/

package main

import (
	"fmt"
	"math"
)

const lim = 7654321

func main() {
	var result = find_pandigital_prime(lim)

	fmt.Println("\t== Project Euler: Problem 41 ==\n")
	fmt.Printf("answer: %d", result)
}

func find_pandigital_prime(limit int) int {
	sieve := make([]bool, limit+1, 2*(limit+1))

	//initialize all values to true
	for i := range sieve {
		sieve[i] = true
	}

	//setting up sieve
	sieve[0] = false
	sieve[1] = false

	//set seive limit
	sieve_lim := math.Sqrt(float64(limit))

	//mark off all multiples
	for i := 0; i < int(sieve_lim); i++ {
		if sieve[i] == true {
			mark_multiple(sieve, i)
		}
	}

	//find and return biggest pandigital prime
	for i := len(sieve) - 1; i != 0; i-- {
		if sieve[i] == true && is_pandigital(i) {
			return i
		}
	}

	return -1
}

//function marks off multiples of the bse passed in
func mark_multiple(array []bool, base int) {
	for i := 2; base*i < len(array); i++ {
		array[base*i] = false
	}
}

//checks if a number is pandigital
func is_pandigital(num int) bool {
	prev, curr := 0, 0

	num_size := int(math.Log10(float64(num))) + 1

	for num != 0 {
		curr = prev | (1 << uint(num%10-1))
		if prev == curr || (num%10) > num_size {
			return false
		}
		num = num / 10
		prev = curr
	}

	return true
}
